### ISSUE TITLE
[Question] Duplicate resource timings on webkit

### ISSUE THREAD

**Issue by Smrtnyk** [microsoft/playwright#18855]
This works on 1.25 for webkit but fails for versions after it.

I have already asked about this before, but answer was vague and no further response was given.
What I am asking here is a bit of clarification, and to know if this can be fixed or not, so my team can move forward (either exclude tests on webkit or try some other package like puppeteer or cypress) since we are already 3 versions behind on playwright.

I have created this page.
```html
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
    <script src="/main.js" ></script>
</head>
<body>
    <img src="https://via.placeholder.com/150" alt="" />
</body>
</html>
```
I served it using vscode live server

and I have this test
```typescript
 test.only("google interceptor", async ({ page }) => {
        await page.route(/google/, () => {
            throw new Error("should never come here!");
        });
        await page.goto("http://127.0.0.1:5500/index.html");
        console.log(await page.evaluate(() => {
            return performance.getEntriesByType("resource");
        }));
    });
```
which I run using `@playwright/test` package.
Basically what happens is that it yields duplicate entries for image in performance.getEntries().

case 1: no route handler for /google/ and main.js on page -> no duplicate entries in resource timings for image
case 2: route handler for /google/ registered, no main.js in html -> no duplicate entries in resource timings
case 3: no route handler for /google/ registered, no main.js in html -> no duplicate entries in resource timings
case 4: route handler for /google./ registered, main.js in html -> duplicate entries in resource timings for image

there is not a single resource on the page that loads something from google, and main.js is completely empty (or it doesn't even exist, doesn't matter)  so there is nothing that would trigger duplicate load.
From the test case above there is not even anything intercepted, because only route handler registered for `google` throws an error so it never comes to that line at all.

I have tests where I need to have the route handlers, and I have scripts on the page, I can't change that, and those work well on firefox and chrome and worked well on previous playwright versions(1.25 ) webkit.

I got a response that it is a preload request, I tried adding preload="none" to image tags and it still doesn't help.
I am just trying to understand how to fix this, because it is breaking lots of our tests now.

Basically what I don't understand here is what is the right behavior?
Is webkit wrong to not report duplicate entries in case there is no route handler registered?
Or is playwright wrong to exhibit such behavior of webkit when there is route handler registered and an empty script on the page?
It looks inconsistent

Does anyone have an idea on how to avoid these duplicate resource entries?
Can we update playwright but somehow still run older webkit which does not show this behavior?

**Comment by pavelfeldman** [comment]
Resource timing is implemented in a platform-specific manner in the WebKit family of products, it basically comes from the respective network stack used. What platform are you using? It might be that request interception and network timing implementation are clashing somewhere. We'll probably look at it with a low priority though. Did you try using https://playwright.dev/docs/next/api/class-request#request-timing? Does it exhibit the same behavior? I would hope since it is lower-level, it would not have dupe entries.

**Comment by Smrtnyk** [comment]
The thing is our test case is bit more complex and I will give some context.

What we have is a tomcat server with jsp pages and test logic is executed using our custom framework we wrote.
We use playwright to visit those pages, wait for test to finish and then collect results.
We don't visit all of the test pages but some of most critical ones so we have fast feedback in case something is broken.
So playwright is like a smoke tests guardian for us which runs on each push to the CI.

But, we don't have the actual test logic in playwright implemented, so we cannot influence the test page itself.
So https://playwright.dev/docs/next/api/class-request#request-timing does not work for us.
Playwright just visits those pages, waits for test to execute and collects results for us to analyze.

These tests also run on dedicated machines, on real SAFARI 14, 15 and 16 and on those machines we don't see such behavior.
Timings are correct and those tests run once a day since these machines run much more tests than we do in playwright.

**Comment by Smrtnyk** [comment]
In the issue I opened previously but was closed
there was an answer that it was a regression introduced by this PR https://github.com/microsoft/playwright-browsers/pull/12 which I cannot see, probably lack of permissions
and then the final answer was that this duplication is correct since it  is a preload request
```text
Well, the new behavior is actually correct. The browser sends two requests for the image - one preload and the other is normal load request. Playwright intercepts both and the timing information is also available for both requests. Previously (in playwright <= v1.25.0) we didn't intercept preload requests.
```
we don't preload anything in our pages, and real safari 16 also doesn't so I am very confused here
I have explained above in the first post on what change to do to get rid of duplication, but it makes no sense to me

**Comment by Smrtnyk** [comment]
> We'll probably look at it with a low priority though

I am fine if this issue will get no priority, we just want to know if at least this will be tackled at some point, or if it is something that should be fixed on your side.
Our product is about real user monitoring and it is very important for us to always test on latest browsers, before we release to customers to be sure everything is working properly.
So staying stuck with older browsers is not an option we can afford, therefore we just want to know if we should look for an alternative, or just give up on testing on webkit, or at least test stuff that don't involve resource timings.

**Comment by pavelfeldman** [comment]
> But, we don't have the actual test logic in playwright implemented, so we cannot influence the test page itself.
So https://playwright.dev/docs/next/api/class-request#request-timing does not work for us.
Playwright just visits those pages, waits for test to execute and collects results for us to analyze.

If Playwright visits those pages and waits for test to execute, then it can report on all the requests along with their timing data. Could you elaborate on why this does not work?

**Comment by Smrtnyk** [comment]
Our real user monitoring product is injected in those test pages and it relies on native browser timings that it gets from performance object, and we are testing that it has proper outcome.
This is how our visiting logic in playwright looks like
```typescript
test("visually complete image size", ({ dtPage }) => {
     return dtPage
          .setFeatureHash("ICAfuN2rV")
          .runTest(BASE_URL + "resource-timing-vc/image-size.jsp");
 });
```
And `runTest` looks like this
```typescript
async runTest(testPath: string): Promise<void> {
        this.testPath = testPath;
        let failures: string[];
        try {
            await this.goto(testPath);
            await this.page.waitForSelector("#pw-finished", { state: "attached", timeout: 30000 });
            failures = await this.page.evaluate(() => {
                return _assert.getCurrentAssertionFailures();
            });
        } catch (e) {
            throw new Error(`Test ${testPath} failed to run because of ${e.message}`);
        }
        if (failures.length) {
            const testFailSummary = `Test ${testPath} failed because of the following errors:\n\t${failures.join("\n\t")}`;
            this.failureMessage = truncateLogMessage(testFailSummary);
            throw new Error(testFailSummary);
        }
    }
```
Basically we visit the page with playwright, wait for invisible DOM element to appear, that is a signal that test executed and report is available.
Then we retrieve the test report from browser context with `_assert.getCurrentAssertionFailures()`
in case there are failures we throw the exception in playwright context.
And this runs in our jenkins CI, we just use playwright to execute important tests.
When we commit to our real user monitoring product, we get feedback using playwright.

We have our custom written java based test framework that runs on real machines and browsers like last 3 chrome versions, last 3 safari versions, last 3 ff versions etc...
Safari runs on mac machines, and there we don't see duplicate resource timings.
And it runs separate from our CI, there are like 80k tests executed and it takes whole day.
Those machines share these test pages, with playwright, where playwright just is used to visit a subset of these tests.

**Comment by pavelfeldman** [comment]
I see. We are a lab tool, so making sure rum tools work in our setting is somewhat a low priority. More importantly, in the WebKit family, you can't collect performance data on non-macOS. While the correctness of execution is guaranteed on Linux and Windows, the network stack there is entirely different from Safari's. And that is by far the largest contributor into the perf numbers you are collecting. So I would start with switching to macOS for meaningful performance numbers first. And as a side-effect, I don't think you will see these dupes there.

**Comment by Smrtnyk** [comment]
Unfortunately our ci in jenkins runs on ubuntu docker images. 
We dont see duplicates in real safari.
So you say that this behavior after playwright 1.25 was not introduced by playwright, but it is a behavior introduced by webkit update?
If so we will then need to stop testing on webkit then.
Or is there a way we can update playwright to still have latest ff and chromium and somehow keep this older webkit from playwright 1.25?

**Comment by pavelfeldman** [comment]
> Or is there a way we can update playwright to still have latest ff and chromium and somehow keep this older webkit from playwright 1.25?

No, that would not be supported.

Let me schedule this for the next milestone, it might be a sign of a bigger issue on Linux and it would be good for us to look into it.

**Comment by yury-s** [comment]
For the record, this is the original bug report where prefetch request caused duplicate entries: https://github.com/microsoft/playwright/issues/17811#issuecomment-1276854980

**Comment by yury-s** [comment]
Debug notes: it seems that after enabling prefetch request interception in WebKit, there is now an extra request is sent / logged for the image url when request is intercepted via page.route.

**Comment by Smrtnyk** [comment]
@aslushnikov seems like this got fixed in 1.32 
I reenabled our tests on webkit and they work properly again

**Comment by aslushnikov** [comment]
Closing then! Thanks for coming back!
